In [52]:
# 1.
with open('quran-simple.txt', 'r', encoding='utf-8') as file:
    quran_text = file.read()
print(len(quran_text))  # print number of characters in the Qur’an text

689411


In [68]:
# 2.
import nltk

arabic_words = nltk.word_tokenize(quran_text)
print(len(arabic_words))  # print number of tokens in the Qur’an

78245


In [54]:
# 3.
unique_arabic_words = set(arabic_words)
print(len(unique_arabic_words))  # print number of unique words in the Qur’an

17574


In [55]:
# 4.
def lexical_diversity(text):
    return len(set(text)) / len(text)

arabic_ld = lexical_diversity(arabic_words)
print(arabic_ld)  # print the lexical diversity of the Qur’an’s vocabulary

0.22460221100389802


In [56]:
# 5.
from nltk.probability import FreqDist

fdist_arabic = FreqDist(arabic_words)
with open('quran_frequency_distribution.txt', 'w', encoding='utf-8') as file:
    for word, frequency in fdist_arabic.most_common(10):
        file.write(f"{word}: {frequency}\n")
print(fdist_arabic.most_common(10))  # print the top 10 words and their frequency

[('مِنْ', 1673), ('فِي', 1185), ('مَا', 1010), ('اللَّهِ', 940), ('لَا', 812), ('الَّذِينَ', 810), ('اللَّهُ', 733), ('مِنَ', 693), ('عَلَى', 670), ('إِلَّا', 662)]


In [77]:
# 6. Read the English text file and store the text in a variable
with open('en.pickthall.txt', 'r', encoding='utf-8') as file:
    English_quran_text = file.read()
print(len(English_quran_text))  # print number of characters in the Qur’an text

828439


In [78]:
# Tokenize the English text, store the tokens in a list, and do the following:
English_words = nltk.word_tokenize(English_quran_text)
print(len(English_words))  # print number of tokens in the Qur’an

183536


In [79]:
# Compute the number of unique words in the English text
unique_English_words = set(English_words)
print(len(unique_English_words))

6849


In [82]:
# Compute the number of tokens (words) in the English text
print(len(English_words))

183536


In [81]:
# Compute the lexical diversity of the English text
English_Id = lexical_diversity(English_words)
print(English_Id)

0.037316929648679276


In [83]:
# Compute the frequency distribution of the English text
fdist_English = FreqDist(English_words)
with open('English_quran_frequency_distribution.txt', 'w', encoding='utf-8') as file:
    for word, frequency in fdist_English.most_common(10):
        file.write(f"{word}: {frequency}\n")
print(fdist_English.most_common(10))

[(',', 8213), ('.', 7470), ('the', 7379), ('and', 5546), ('of', 4838), ('is', 3095), (')', 2974), ('(', 2973), ('Allah', 2914), ('they', 2527)]


7.

Comparing the results, we can see that:

The Quran text in Arabic has less characters than the English translation of the Quran, with 689411 characters in Arabic compared to 828439 characters in English.

The number of tokens in Arabic is also less than in English, with 78245 tokens in Arabic compared to 183536 tokens in English.


The number of unique words in Arabic is higher than in English, with 17574 unique words in Arabic compared to 6849 unique words in English.


The lexical diversity of the Quran's vocabulary is less in English than in Arabic, with a lexical diversity of 0.037 in English compared to 0.224 in Arabic. This means that the vocabulary in the Arabic Quran is less varied than in the English translation.


The frequency distribution of the top 10 words is also different in Arabic and English, with "مِنْ" (and) being the most frequent word in Arabic and "the" being the most frequent word in English.

In [85]:
import re

# 8. Define a function to remove short vowels and diacritics from Arabic words
def remove_short_vowels_and_diacritics(word):
    """
    A function that removes short vowels and diacritics from an Arabic word.
    """
    # Define a regular expression pattern to match all diacritic characters
    pattern = re.compile('[\u064B-\u0652]')
    
    # Use the pattern to remove all diacritic characters from the word
    non_vowelized_word = re.sub(pattern, '', word)
    
    return non_vowelized_word

# 9. Read the Arabic Quran text file and store the text in a variable
with open('quran-simple.txt', 'r', encoding='utf8') as f:
    quran_text = f.read()

# Tokenize the Quran text into words and store the words in a list
import nltk
nltk.download('punkt')
quran_words = nltk.word_tokenize(quran_text)

# Convert the list of words to a set to get the number of unique words
quran_unique_words = set(quran_words)

# Calculate the number of unique words and the lexical diversity
quran_unique_word_count = len(quran_unique_words)
quran_lexical_diversity = len(quran_words) / quran_unique_word_count

# Compute the frequency distribution of words in the Quran and print the top 10 words
from nltk import FreqDist
quran_freq_dist = FreqDist(quran_words)
print(quran_freq_dist.most_common(10))

# Generate a list of non-vowelized forms for all words in the Quran
quran_non_vowelized = [remove_short_vowels_and_diacritics(word) for word in quran_words]

# Convert the list of non-vowelized forms to a set to get the number of unique words
quran_unique_non_vowelized = set(quran_non_vowelized)

# Calculate the number of unique words and the lexical diversity of the non-vowelized forms
quran_unique_non_vowelized_count = len(quran_unique_non_vowelized)
quran_lexical_diversity_non_vowelized = len(quran_non_vowelized) / quran_unique_non_vowelized_count

# Compute the frequency distribution of non-vowelized words in the Quran and print the top 10 words
quran_freq_dist_non_vowelized = FreqDist(quran_non_vowelized)
print(quran_freq_dist_non_vowelized.most_common(10))

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


[('مِنْ', 1673), ('فِي', 1185), ('مَا', 1010), ('اللَّهِ', 940), ('لَا', 812), ('الَّذِينَ', 810), ('اللَّهُ', 733), ('مِنَ', 693), ('عَلَى', 670), ('إِلَّا', 662)]
[('من', 2763), ('الله', 2265), ('في', 1185), ('ما', 1010), ('إن', 966), ('لا', 812), ('الذين', 810), ('على', 670), ('إلا', 664), ('ولا', 658)]
